hazelcast/src/main/java/com/hazelcast/durableexecutor/impl/DistributedDurableExecutorService.java

```diff
@@ -0,0 +1,132 @@
+        partitionContainers = new DurablePartitionContainer[partitionCount];
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+            partitionContainers[partitionId] = new DurablePartitionContainer(nodeEngine, partitionId);
+        }
+    }
+
+    @Override
+    public void init(NodeEngine nodeEngine, Properties properties) {
+    }
+
+    public DurablePartitionContainer getPartitionContainer(int partitionId) {
+        return partitionContainers[partitionId];
+    }
+
+    public NodeEngine getNodeEngine() {
+        return nodeEngine;
+    }
+
+    @Override
+    public void reset() {
+        shutdownExecutors.clear();
+        for (int partitionId = 0; partitionId < partitionContainers.length; partitionId++) {
+            partitionContainers[partitionId] = new DurablePartitionContainer(nodeEngine, partitionId);
+        }
+    }
+
+    @Override
+    public void shutdown(boolean terminate) {
+        reset();
+    }
+
+    @Override
+    public DistributedObject createDistributedObject(String name) {
+        return new DurableExecutorServiceProxy(nodeEngine, this, name);
+    }
+
+    @Override
+    public void destroyDistributedObject(String name) {
+        shutdownExecutors.remove(name);
+        nodeEngine.getExecutionService().shutdownExecutor(name);
+    }
+
+    public void shutdownExecutor(String name) {
+        nodeEngine.getExecutionService().shutdownExecutor(name);
+        shutdownExecutors.add(name);
+    }
+
+    public boolean isShutdown(String name) {
+        return shutdownExecutors.contains(name);
+    }
+
+    @Override
+    public Operation prepareReplicationOperation(PartitionReplicationEvent event) {
+        int partitionId = event.getPartitionId();
+        DurablePartitionContainer partitionContainer = partitionContainers[partitionId];
+        return partitionContainer.prepareReplicationOperation(event.getReplicaIndex());
+    }
+
+    @Override
+    public void beforeMigration(PartitionMigrationEvent event) {
+    }
+
+    @Override
+    public void commitMigration(PartitionMigrationEvent event) {
+        int partitionId = event.getPartitionId();
+        if (event.getMigrationEndpoint() == MigrationEndpoint.SOURCE) {
+            clearMapsHavingLesserBackupCountThan(partitionId, event.getNewReplicaIndex());
+        } else if (event.getNewReplicaIndex() == 0) {
+            DurablePartitionContainer partitionContainer = partitionContainers[partitionId];
+            partitionContainer.executeAll();
+        }
+    }
+
+    @Override
+    public void rollbackMigration(PartitionMigrationEvent event) {
+        if (event.getMigrationEndpoint() == MigrationEndpoint.DESTINATION) {
+            clearMapsHavingLesserBackupCountThan(event.getPartitionId(), event.getCurrentReplicaIndex());
+        }
+    }
+
+    private void clearMapsHavingLesserBackupCountThan(int partitionId, int thresholdReplicaIndex) {
```

Comment:
minor comment: I guess method name should be `clearRingBuffersHavingLesserBackupCountThan`